### PR TITLE
Feature/706_plugin mecanism_with_pluggy

### PR DESCRIPTION
**Pull request by inkhey**
Related to https://github.com/tracim/tracim/issues/706 and https://github.com/tracim/tracim/issues/1487

## Checkpoints

The goal of this section is to help code reviewer to do required checks on this pull request. Please don't edit this list.
If one or more of these checkpoints are out of scope, please write below why they are.

- [x] Code is clear enough
- [x] If there are FIXME in the code, then there are associated issues and issue is referenced in the FIXME
- [x] If there are TODO, NOTE or HACK in code, date and developer initials are present
- [x] Automated tests have been written (covering feature or non regression), or *at least* an issue has been created for test implementation

### CODE REVIEW

File: backend/requirements.txt [hunk 1]
```diff
@@ -100,6 +103,10 @@ webencodings==0.5.1       # via tinycss2
 webob==1.8.4              # via pyramid
 wsgidav==2.4.1
 xvfbwrapper==0.2.9        # via preview-generator
+zipp==0.6.0               # via importlib-metadata
 zope.deprecation==4.3.0   # via pyramid
 zope.interface==4.6.0     # via pyramid, pyramid-retry, transaction, zope.sqlalchemy
 zope.sqlalchemy==1.0
+
+# The following packages are considered to be unsafe in a requirements file:
+# setuptools
```
Comment (buxx) [inline backend/requirements.txt]: This is generated, but cant it be skipped easly ?

File: backend/tracim_backend/lib/core/plugins.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+HOOKSPEC_FILES_SUFFIX = "hookspec"
+hookspec = pluggy.HookspecMarker(EVENT_NAMESPACE)
+hookimpl = pluggy.HookimplMarker(EVENT_NAMESPACE)
+
+
+class TracimPluginManager(object):
+    def __init__(self, prefix: str):
+        self.prefix = prefix
+        self.loaded_plugin = False
+        self.plugins = {}
```
Comment (buxx) [inline backend/tracim_backend/lib/core/plugins.py]: typing

File: backend/tracim_backend/lib/core/plugins.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        root_module = importlib.import_module(root_module_name)
+        # INFO - G.M - 2019-11-28 - get all submodules recursively,
+        # find those with with hookspec suffix, import them and add them
+        # to valid hookspec
+        for module_path in find_all_submodule_path(root_module):
+            if module_path.endswith(HOOKSPEC_FILES_SUFFIX):
+                hookspec_module_paths.append(module_path)
+        return hookspec_module_paths
+
+    def _load_spec(self, event_dispatcher: pluggy.PluginManager) -> None:
```
Comment (buxx) [inline backend/tracim_backend/lib/core/plugins.py]: `event_dispatcher` is that the correct var name ?

File: backend/tracim_backend/lib/core/plugins.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+class TracimPluginManager(object):
+    def __init__(self, prefix: str):
+        self.prefix = prefix
+        self.loaded_plugin = False
+        self.plugins = {}
+        self.event_dispatcher = self._setup_pluggy_event_dispatcher()
+
+    @classmethod
+    def get_all_hookspec_module_path(cls):
```
Comment (buxx) [inline backend/tracim_backend/lib/core/plugins.py]: typing

File: backend/tracim_backend/lib/core/plugins.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        """
+
+        # INFO - G.M - 2019-11-28 - get all submodules recursively,
+        # find those with with hookspec suffix, import them and add them
+        # to valid hookspec
+        for hookspec_module_path in self.get_all_hookspec_module_path():
+            module = importlib.import_module(hookspec_module_path)
+            event_dispatcher.add_hookspecs(module)
+
+    def _setup_pluggy_event_dispatcher(self) -> pluggy.PluginManager:
```
Comment (buxx) [inline backend/tracim_backend/lib/core/plugins.py]: Prefer `create` naming for creation purpose

File: backend/tracim_backend/lib/core/plugins.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        for module_name, module in sorted_plugins.items():
+            self.register(module)
+
+    def add_plugin_path(self, plugin_folder: str) -> None:
+        """
+        Add path for plugin search
+        :param plugin_folder: folder path (should be absolute path) where plugin should be stored
+        :return:
+        """
+        sys.path.append(plugin_folder)
```
Comment (buxx) [inline backend/tracim_backend/lib/core/plugins.py]: Maybe add it only if not in it ? To avoid mistakes.

File: backend/requirements.txt (hunk 1 above)
Comment (inkhey) [reply backend/requirements.txt]: Don't think so, it'seem that one of our dependencies require setuptools.

Comment (inkhey) [reply backend/requirements.txt]: Checked. Don't see way to disable this line. we can allow unsafe requirement but that's not a good idea.
Just done it to know which dependencies require setuptools:
```
# The following packages are considered to be unsafe in a requirements file:
setuptools==42.0.2        # via cairocffi, plaster, pyramid, tinycss2, zope.deprecation, zope.interface, zope.sqlalchemy

```

File: backend/tracim_backend/lib/core/plugins.py (hunk 4 above)
Comment (inkhey) [reply backend/tracim_backend/lib/core/plugins.py]: done.

File: backend/tracim_backend/lib/core/plugins.py (hunk 2 above)
Comment (inkhey) [reply backend/tracim_backend/lib/core/plugins.py]: done.

File: backend/tracim_backend/lib/core/plugins.py (hunk 6 above)
Comment (inkhey) [reply backend/tracim_backend/lib/core/plugins.py]: done.

File: backend/tracim_backend/lib/core/plugins.py (hunk 5 above)
Comment (inkhey) [reply backend/tracim_backend/lib/core/plugins.py]: done.

**Pull request merged**